probe-rs-cli-util/src/common_options.rs

```diff
@@ -0,0 +1,453 @@
+        };
+
+        Ok(target)
+    }
+
+    /// Attaches to specified probe and configures it.
+    pub fn attach_probe(&self) -> Result<Probe, OperationError> {
+        let mut probe = {
+            if self.dry_run {
+                Probe::from_specific_probe(Box::new(FakeProbe::new()));
+            }
+
+            // If we got a probe selector as an argument, open the probe
+            // matching the selector if possible.
+            match &self.probe_selector {
+                Some(selector) => {
+                    Probe::open(selector.clone()).map_err(OperationError::FailedToOpenProbe)
+                }
+                None => {
+                    // Only automatically select a probe if there is
+                    // only a single probe detected.
+                    let list = Probe::list_all();
+                    if list.len() > 1 {
+                        return Err(OperationError::MultipleProbesFound { number: list.len() });
+                    }
+
+                    if let Some(info) = list.first() {
+                        Probe::open(info).map_err(OperationError::FailedToOpenProbe)
+                    } else {
+                        Err(OperationError::NoProbesFound)
+                    }
+                }
+            }
+        }?;
+
+        // Select protocol and speed
+        probe.select_protocol(self.protocol).map_err(|error| {
+            OperationError::FailedToSelectProtocol {
+                source: error,
+                protocol: self.protocol,
+            }
+        })?;
+        if let Some(speed) = self.speed {
+            let _actual_speed = probe.set_speed(speed).map_err(|error| {
+                OperationError::FailedToSelectProtocolSpeed {
+                    source: error,
+                    speed,
+                }
+            })?;
+        }
+
+        Ok(probe)
+    }
+
+    pub fn attach_session(
+        &self,
+        probe: Probe,
+        target: TargetSelector,
+    ) -> Result<Session, OperationError> {
+        let session = if self.connect_under_reset {
+            probe.attach_under_reset(target)
+        } else {
+            probe.attach(target)
+        }
+        .map_err(|error| OperationError::AttachingFailed {
+            source: error,
+            connect_under_reset: self.connect_under_reset,
+        })?;
+
+        Ok(session)
+    }
+
+    /// Convenience method that attaches to the specified probe, target,
+    /// and target session.
+    pub fn create_session(&self) -> Result<Session, OperationError> {
+        let target = self.get_target_selector()?;
+        let probe = self.attach_probe()?;
+        let session = self.attach_session(probe, target)?;
+
+        Ok(session)
+    }
+
+    /// Builds a new flash loader for the given target and ELF. This
+    /// will check the ELF for validity and check what pages have to be
+    /// flashed etc.
+    pub fn build_flashloader(
+        &self,
+        target: &Option<TargetSelector>,
```

Comment:
I'm not wholly accustomed to the API. I presumed there are cases when a target cannot be established from a session, so that the user must decide (this function was copied verbatim; I have only applied small changes). If this is not the case I'll simplify.

Reply:
In the original function the `Session` was not available, so the target argument is necessary. In this case here, just having the `Session` is enough.

(I assume this is the original function: https://github.com/probe-rs/cargo-flash/blob/4daa1227e98bc78e748373f5f6eb9c7ffc04bd1a/src/main.rs#L724)